Add gas recyclers

Partmedia opened the pull request:
## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->
Gas recyclers catalyze the conversion of carbon dioxide into oxygen and nitrous oxide into nitrogen. The gas recycler component takes gas from the input net, converts carbon dioxide to oxygen, and releases the result into the output net.

To make things more fun, the input net must be pressurized to 3 MPa and heated to at least 300 C; otherwise, no reaction will occur. It takes some atmospheric knowledge to set up such an environment, but one way this can be achieved is by using (see screenshots):

1. An input pump to bring the pressure of a "small enough" input pipe net beyond 3 MPa,
2. A heater to bring the input pipe net up to temperature,
3. The gas recycler itself,
4. A freezer to prevent the recycled gases from burning the station inhabitants, and
5. Some output net flow rate limiter, so that the input pressure and temperature stay high enough.

Game-mechanic wise, gas recyclers contain the catalyst for the conversion reaction, and therefore, requires no external power. However, the external pumps and heaters required to make the reaction happen still do.

Thank you to @Peptide90 for the sprite and prototype updates.

**Screenshots**
![gas_recycler_3_annotated](https://user-images.githubusercontent.com/3229565/183011503-582c0878-d235-4fab-8a06-7de0ea177dec.png)

**Changelog**
:cl: notafet
- add: Gas recyclers can now convert waste carbon dioxide and nitrous oxide into breathable gases.

ike709 commented:
Would be nice if it required another gas like tritium.

juliangiebel commented:
Maybe it should require you to replace the catalyst once in a while and each conversion takes a different catalyst.

juliangiebel reviewed (commented):
Good PR :smile:
I only found that you named the parameter for component in your event handlers `c` and single letter variables should be avoided imo.
Otherwise the code looks good. :+1:
I still have to check the PR out in game when I get home.

Partmedia commented on Content.Server/Atmos/Piping/Binary/EntitySystems/GasRecyclerSystem.cs:
```diff
@@ -0,0 +1,111 @@
+            UpdateAppearance(uid, c);
+        }
+
+        private void OnExamined(EntityUid uid, GasRecyclerComponent c, ExaminedEvent args)
```
I'll make the change.

Partmedia commented on Content.Server/Atmos/Piping/Binary/EntitySystems/GasRecyclerSystem.cs:
```diff
@@ -0,0 +1,111 @@
+
+            if (inlet.Air.Temperature >= minTemp && inlet.Air.Pressure >= minPressure)
+            {
+                c.Reacting = true;
+                var nCO2 = inlet.Air.GetMoles(Gas.CarbonDioxide);
+                inlet.Air.AdjustMoles(Gas.CarbonDioxide, -nCO2);
+                inlet.Air.AdjustMoles(Gas.Oxygen, nCO2);
+                var nN2O = inlet.Air.GetMoles(Gas.NitrousOxide);
+                inlet.Air.AdjustMoles(Gas.NitrousOxide, -nN2O);
+                inlet.Air.AdjustMoles(Gas.Nitrogen, nN2O);
```
The original intent of this was to convert only the fraction of gas that passes through the device. That bit of calculation is in my passive gate PR (not yet merged when I wrote this) and I didn't want to duplicate code.

I would need to move some common code from PassiveGateSystem to AtmosphereSystem and use it here. Would you be willing to review that in this PR or a separate one?

juliangiebel replied:
It might be better to do that in a separate PR so this one doesn't have to wait on your passive gate PR.

Partmedia commented:
> Maybe it should require you to replace the catalyst once in a while.

Catalysts [aren't generally consumed](https://en.wikipedia.org/wiki/Catalytic_converter). But it would be fun to damage them if someone tries to flow toxic gases through the recycler.

I don't want to make using recyclers too hard right now, though, especially since we have gas miners on all stations.

juliangiebel commented:
> > Maybe it should require you to replace the catalyst once in a while.
> 
> Catalysts [aren't generally consumed](https://en.wikipedia.org/wiki/Catalytic_converter). But it would be fun to damage them if someone tries to flow toxic gases through the recycler.
> 
> I don't want to make using recyclers too hard right now, though, especially since we have gas miners on all stations.

You're right. We can change things later on when gas miners get replaced / removed.

mirrorcult commented on Content.Server/Atmos/Piping/Binary/EntitySystems/GasRecyclerSystem.cs:
```diff
@@ -0,0 +1,111 @@
+                return;
+            }
+
+            if (comp.Reacting && Loc.TryGetString("gas-recycler-reacting", out var str))
+            {
+                args.PushMarkup(str);
+            }
```
this is kind of a weird pattern, most everywhere else in the codebase would just do this

```suggestion
            if (comp.Reacting)
            {
                args.PushMarkup(Loc.GetString("gas-recycler-reacting"));
            }
```

etc for the others

mirrorcult commented on Content.Server/Atmos/Piping/Binary/EntitySystems/GasRecyclerSystem.cs:
```diff
@@ -0,0 +1,111 @@
+            SubscribeLocalEvent<GasRecyclerComponent, ExaminedEvent>(OnExamined);
+        }
+
+        private void OnInit(EntityUid uid, GasRecyclerComponent comp, ComponentInit args)
+        {
+            UpdateAppearance(uid, comp);
+        }
```
shouldnt this be on `AtmosDeviceEnabledEvent` rather than init (also as a counterpart to the disabled event being handled)

mirrorcult commented on Content.Server/Atmos/Piping/Binary/EntitySystems/GasRecyclerSystem.cs:
```diff
@@ -0,0 +1,111 @@
+        [Dependency] private readonly AtmosphereSystem _atmosphereSystem = default!;
+        [Dependency] private readonly SharedAmbientSoundSystem _ambientSoundSystem = default!;
+
+        private const float minTemp = 300 + 273; // 300 C
+        private const float minPressure = 3000;  // 3 MPa
```
```suggestion
        private const float MinTemp = 300 + Atmospherics.T0C; // 300 C
        private const float MinPressure = 30 * Atmospherics.OneAtmosphere;  // 3 MPa
```

c# convention + define this in terms of other atmos constants

mirrorcult reviewed (changes requested):
concept looks good to me. with this i'd be in favor of removing miners entirely from all maps and making the tanks simply decently pressurized at roundstart

mirrorcult commented:
passive gate PR is also merged now so might want to integrate the gradual change in here?

Partmedia commented:
Sorry for the rebase/force-push; I needed to resync with upstream master.

Changes:

- Addressed reviewer feedback
- Implement gradual gas conversion. I ended up not using code from the passive gate to make it "feel" better in game. This involved adding a `RemoveVolume()` method, which is a separate change in a separate commit that should probably be left separate (i.e. rebase only, instead of squash-then-rebase).

I was able to convert atmos on a mid-sized map to work without miners and set up the gas recycler so that it would convert waste gases. Before removing miners anywhere though, I'd suggest having a bit of a transition period so I can write up what mappers need to add on the wiki, and see if any of the constants need some tweaking.

Also taking a look at the failing build right now.

Partmedia commented:
Fixed integration tests; fixed construction recipe.

Peptide90 commented:
Very odd that tests are failing due to a map file that isn't touched. Probably fine to merge.

Partmedia commented:
Thanks! For those following space exploration in the real world:

> MOXIE can reliably convert carbon dioxide into a small tree’s worth of oxygen... MOXIE works its magic by sucking in air, filtering out dust, and compressing and heating the gases to 800 degrees Celsius. The heated air flows through a solid oxide electrolysis instrument that splits carbon dioxide... into oxygen and carbon monoxide... then separates out the oxygen and expels the carbon monoxide, alongside other gases, as exhaust.

_Source: [NASA Can Now Reliably Produce a Tree’s Worth of Oxygen on Mars](https://singularityhub.com/2022/09/04/nasa-says-it-can-now-reliably-produce-a-trees-worth-of-oxygen-on-mars/), Sept. 4, 2022_
